src/timer.rs

```diff
@@ -77,3 +77,17 @@ pub trait CountDown {
 
 /// Marker trait that indicates that a timer is periodic
 pub trait Periodic {}
+
+/// Trait for cancelable countdowns.
+pub trait Cancel : CountDown {
+    /// Error returned when a countdown can't be canceled.
+    type Error;
+
+    /// Tries to cancel this countdown.
+    ///
+    /// # Errors
+    ///
+    /// An error will be returned if the countdown has already been canceled or was never started.
+    /// An error is also returned if the countdown is not `Periodic` and has already expired.
```

Comment:
I'm wondering if the documentation should be that specific. Maybe setting `Error` to `!` is a valid thing for an implementation to decide.

I'm unsure. Unless others have stronger opinions, I'd feel more comfortable if the documentation were less definitive ("Possible error conditions include ...", something like that).

Reply:
I think what you say is the right approach for *most* traits `embedded-hal` defines, but not this one: It's important for applications to be able to rely on the possible reasons this could fail, because applications will potentially choose to *ignore* cancellation errors deliberately if the timer has done its job and is no longer needed (ie. the application doesn't care if the timer expired in the meantime).

If the error conditions were unclear, applications would have to `.unwrap()` every time to be sure, which can cause sporadic panics if the timer expires shortly before that.

> Maybe setting `Error` to `!` is a valid thing for an implementation to decide.

This comes with another problem: If this is done, applications can no longer use `cancel` to determine if the timer has expired before the call. @therealprof has mentioned in https://github.com/japaric/embedded-hal/pull/67#issuecomment-380462837 that this might be used by real-time systems to determine if a deadline was missed, which seems like a very valid use case to me (and in fact the only use case that convinced me that returning a `Result` is useful at all).

So yeah, I think defining the error conditions to *some* extent is required for this to be useful at all.

Reply:
Thank you, that makes a lot of sense. In that case we should consider whether pre-defining an error enum (instead of having the associated type) would be a good idea. In the interest of moving things along, I consider that a question for later though. I'll open follow-up issues tomorrow and will make sure to mention that.